static/panes/selectiondag-view.js

```diff
@@ -0,0 +1,567 @@
+            '[<!ENTITY nbsp "&#160;">]>';
+        var name = this.currentFunc + "-" + this.currentDag;
+        var graphSvg = this.placeholder.select("svg");
+        var svgContents = graphSvg 
+            .attr("title", name)
+            .attr("version", 1.1)
+            .attr("xmlns", "http://www.w3.org/2000/svg")
+            .attr("xmlns:xhtml", "http://www.w3.org/1999/xhtml")
+            .attr("xmlns:xlink", "http-www.w3.org/1999/xlink")
+            .node().parentNode.innerHTML;
+
+        saveAs(new Blob(
+            [doctype + svgContents],
+            {type: "image/svg+xml"}),
+        name + ".svg");
+        return true;
+    } catch (e) {
+        this.alertSystem.notify('Error while saving your selection DAG. Use the clipboard instead.', {
+            group: "savelocalerror",
+            alertClass: "notification-error",
+            dismissTime: 5000
+        });
+        return false;
+    }
+};
+
+SelectionDAG.prototype.saveDAGAsPNG = function () {
+    try {
+        var name = this.currentFunc + "-" + this.currentDag;
+        var graphSvg = this.placeholder.select("svg");
+        
+        // let svgContents = graphSvg
+        //     .attr("title", name)
+        //     .attr("version", 1.1)
+        //     .attr("xmlns", "http://www.w3.org/2000/svg")
+        //     .attr("xmlns:xhtml", "http://www.w3.org/1999/xhtml")
+        //     .attr("xmlns:xlink", "http-www.w3.org/1999/xlink")
+        //     .node().parentNode.innerHTML;
+
+        var svgContents = this.getSVGString(graphSvg.node());
+        var width = graphSvg.attr('width').replace('pt', '');
+        var height = graphSvg.attr('height').replace('pt', '');
+        
+        this.SvgToImage(svgContents, 2*width, 2*height, 'png', function (dataBlob) {
+            saveAs(dataBlob, name + ".png");
+        });         
+        return true;
+    } catch (e) {
+        this.alertSystem.notify('Error while saving your selection DAG. Use the clipboard instead.', {
+            group: "savelocalerror",
+            alertClass: "notification-error",
+            dismissTime: 5000
+        });
+        return false;
+    }
+};
+
+SelectionDAG.prototype.SvgToImage =  function (svgString, width, height, fmt, callback ) {
+    var imgSrc = 'data:image/svg+xml;base64,' + 
+        btoa(unescape(encodeURIComponent(svgString))); // Convert SVG string to data URL
+
+    var canvas = document.createElement("canvas");
+    var context = canvas.getContext("2d");
+
+    canvas.width = width;
+    canvas.height = height;
+
+    var image = new Image();
+    image.onload = function () {
+        context.clearRect ( 0, 0, width, height );
+        context.drawImage(image, 0, 0, width, height);
+
+        canvas.toBlob( function (blob) {
+            var fileSize = Math.round(blob.length / 1024) + ' KB';
+            if ( callback ) callback( blob, fileSize );
+        });
+
+
+    };
+
+    image.src = imgSrc;
```

Comment:
This is blocking png export right now.

While I'm assigning the imgSrc to the generates image here, I can't get the `.onload` even to fire for some reason.

I'm sure it's something I don't understand about the environment that is causing this. If someone more knowledgeable about JS and the runtime environment where this is running on the client side, maybe I can help and be nudged into solving this.

Reply:
This sounds less of a JS issue and more an HTML/DOM thing. My guess is the image isn't really "loaded" until it's placed into the document somewhere. I'd suggest tryinmg adding it to the `document`  and then if it fires you're golden. It'll then need to be removed from the document, and we might have to find a way to hide it in some way.

or maybe you need to use `document.createElement('img');` to link it to the document somehow? I know the ways of the DOM are subtle and quick to anger... just some thoughts!